File: glass-app/src/main/java/dev/glass/glass/transport/TransportFactory.java
package dev.glass.glass.transport;

import android.os.Build;
import android.util.Log;

import dev.glass.protocol.transport.TcpTransport;
import dev.glass.protocol.transport.Transport;

import java.lang.reflect.Method;

/**
 * Glass-side transport selector mirroring the phone's {@code TransportFactory}.
 *
 * Resolution: setprop > BuildConfig > auto (emulator → TCP, real device → RFCOMM).
 */
public final class TransportFactory {
    private static final String TAG = "TransportFactory";
    private TransportFactory() {}

    public static Transport create() {
        String kind = resolveKind();
        Log.i(TAG, "creating transport: kind=" + kind);
        switch (kind) {
            case "rfcomm":
                return new RfcommServerTransport();
            case "tcp":
                return new TcpTransport(TcpTransport.Role.SERVER, "0.0.0.0", 8765);
            default:
                throw new IllegalStateException("unknown transport kind: " + kind);
        }
    }

    private static String resolveKind() {
        // Resolution order: setprop override > auto-detect (emulator → TCP, real device → RFCOMM).
        // BuildConfig.TRANSPORT_KIND is ignored: the debug build sets it to "tcp" for emulator
        // pair tests, but on a real Glass we always want RFCOMM unless explicitly overridden.
        String prop = systemProp("dev.glass.transport");
        if (prop != null && !prop.isEmpty()) return prop.toLowerCase();
        return isEmulator() ? "tcp" : "rfcomm";
    }

    private static boolean isEmulator() {
        String product = String.valueOf(Build.PRODUCT).toLowerCase();
        String brand = String.valueOf(Build.BRAND).toLowerCase();
        return product.contains("sdk") || product.contains("emulator")
            || brand.startsWith("generic")
            || String.valueOf(Build.FINGERPRINT).startsWith("generic");
    }

    private static String systemProp(String key) {
        try {
            Class<?> cls = Class.forName("android.os.SystemProperties");
            Method get = cls.getMethod("get", String.class);
            Object v = get.invoke(null, key);
            return v instanceof String ? (String) v : null;
        } catch (Throwable t) {
            return null;
        }
    }
}

File: phone-app/src/main/java/dev/glass/phone/transport/TransportFactory.kt
package dev.glass.phone.transport

import android.content.Context
import android.os.Build
import android.util.Log
import dev.glass.phone.BuildConfig
import dev.glass.protocol.transport.TcpTransport
import dev.glass.protocol.transport.Transport

/**
 * Selects between {@link TcpTransport} (for emulator pair-testing) and {@link RfcommTransport}
 * (real Glass + phone).
 *
 * Resolution order (first match wins):
 *   1. SharedPreferences mode (set via in-app settings or `adb shell setprop dev.glass.transport`)
 *   2. BuildConfig.TRANSPORT_KIND
 *   3. Auto: emulator → TCP, real device with paired Glass → RFCOMM, real device without → TCP
 *
 * Returns a {@link CreateResult} describing what was chosen so the UI can surface diagnostics.
 */
object TransportFactory {

    private const val TAG = "TransportFactory"

    sealed class CreateResult {
        data class Ok(val transport: Transport, val description: String) : CreateResult()
        data class Failed(val reason: String) : CreateResult()
    }

    fun create(ctx: Context): CreateResult {
        val systemMode = systemProp("dev.glass.transport")
        val prefMode = TransportPrefs.getMode(ctx)
        val resolved: TransportPrefs.Mode = when {
            systemMode == "tcp" -> TransportPrefs.Mode.TCP
            systemMode == "rfcomm" -> TransportPrefs.Mode.RFCOMM
            prefMode != TransportPrefs.Mode.AUTO -> prefMode
            isEmulator() -> TransportPrefs.Mode.TCP
            else -> TransportPrefs.Mode.RFCOMM
        }
        Log.i(TAG, "resolved transport mode: $resolved (system=$systemMode, pref=$prefMode, emulator=${isEmulator()})")

        return when (resolved) {
            TransportPrefs.Mode.TCP -> {
                val host = TransportPrefs.getTcpHost(ctx)
                CreateResult.Ok(
                    TcpTransport(TcpTransport.Role.CLIENT, host, 8765),
                    "TCP $host:8765",
                )
            }
            TransportPrefs.Mode.RFCOMM, TransportPrefs.Mode.AUTO -> {
                when (val r = GlassDeviceFinder.find(ctx)) {
                    is GlassDeviceFinder.Result.Found ->
                        CreateResult.Ok(
                            RfcommTransport(r.mac),
                            "RFCOMM ${r.mac} (${r.name}, src=${r.source})",
                        )
                    GlassDeviceFinder.Result.BluetoothUnavailable ->
                        CreateResult.Failed("This device has no Bluetooth.")
                    GlassDeviceFinder.Result.BluetoothDisabled ->
                        CreateResult.Failed("Bluetooth is off — turn it on and reopen the app.")
                    GlassDeviceFinder.Result.PermissionMissing ->
                        CreateResult.Failed("Bluetooth permission not granted.")
                    GlassDeviceFinder.Result.NoPairedDevices ->
                        CreateResult.Failed("Pair the Glass with this phone in Bluetooth settings first.")
                    GlassDeviceFinder.Result.NoGlassFound ->
                        CreateResult.Failed(
                            "Multiple paired devices, none look like Glass — set the MAC in app settings.",
                        )
                }
            }
        }
    }

    private fun isEmulator(): Boolean {
        val product = Build.PRODUCT.lowercase()
        val brand = Build.BRAND.lowercase()
        val device = Build.DEVICE.lowercase()
        val model = Build.MODEL.lowercase()
        return product.contains("sdk")
            || product.contains("emulator")
            || brand.startsWith("generic")
            || device.contains("emulator")
            || model.contains("sdk")
            || Build.FINGERPRINT.startsWith("generic")
    }

    private fun systemProp(key: String): String? = try {
        @Suppress("PrivateApi")
        val cls = Class.forName("android.os.SystemProperties")
        val get = cls.getMethod("get", String::class.java)
        (get.invoke(null, key) as? String)?.takeIf { it.isNotBlank() }
    } catch (t: Throwable) {
        null
    }
}
